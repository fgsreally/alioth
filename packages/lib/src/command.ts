export interface DefaultCommand {
  /** 命令名 */

  name: string
  execute: () => {
    /** 前进时执行的操作 */
    redo?(): void
    /** 撤回时执行的操作 */
    undo?(): void
  }
  /** 键盘 如 shift+y */

  keyboard?: string
  /** 初始化时执行的操作 */
  init?: () => () => void
  /** 是否压入栈，如为否，不参与撤回/前进 */

  pushQueue?: boolean
}

interface commandState<Command> {
  isActive: boolean
  // 索引、指针
  current: number
  // 记录栈
  queue: { redo: any; undo: any }[]
  commands: { [key in string]: () => void } // 命令
  commandArray: Command[]// 所有命令
  destroyArray: Function[]// 销毁任务
}

// 创建指令，主要用于快捷键
export function createCommand<Command extends DefaultCommand>(options: {
  redo?: boolean
  undo?: boolean
} = {}) {
  let initialized = false
  const state: commandState<Command> = {
    isActive: true,
    current: -1,
    queue: [],
    commands: {},
    commandArray: [],
    destroyArray: [],
  }

  const registry = (command: Command) => {
    if(initialized&&command.init){

       state.destroyArray.push(command.init()),

    }
    state.commandArray.push(command)
    state.commands[command.name] = () => {
      // 命令名字对应执行函数
      const { redo, undo } = command.execute() || {}
      if (!command.pushQueue)
        return

      let { queue, current } = state
      if (queue.length > 0) {
        queue = queue.slice(0, current + 1)
        state.queue = queue
      }
      queue.push({
        redo,
        undo,
      })
      state.current = current + 1
    }
  }
  // 注册我们需要的命令
  if (options.redo) {
    registry({
      name: 'redo',
      keyboard: 'ctrl+y',
      execute() {
        const item = state.queue[state.current + 1]
        if (item) {
          item.redo && item.redo()
          state.current++
        }
      },
    } as Command)
  }

  if (options.undo) {
    registry({
      name: 'undo',
      keyboard: 'ctrl+z',
      execute() {
        if (state.current === -1)
          return
        const item = state.queue[state.current]
        if (item) {
          item.undo && item.undo()
          state.current--
        }
      },
    } as Command)
  }

  const keyboardEvent = (() => {
    const onKeydowm = (e: KeyboardEvent) => {
      const { ctrlKey, key, altKey, shiftKey } = e
      // ctrl+z / ctrl+y
      let keyString: any = []
      if (altKey)
        keyString.push('alt')
      if (ctrlKey)
        keyString.push('ctrl')
      if (shiftKey)
        keyString.push('shift')
      keyString.push(key.toLowerCase())

      keyString = keyString.join('+')
      state.commandArray.forEach(({ keyboard, name }) => {
        if (!keyboard)
          return // 没有键盘事件
        if (keyboard === keyString) {
          state.commands[name]()
          e.preventDefault()
        }
      })
    }

    const init = () => {
      // 初始化事件
      window.addEventListener('keydown', onKeydowm)
      return () => {
        // 销毁事件
        window.removeEventListener('keydown', onKeydowm)
      }
    }
    return init
  })()

  state.destroyArray.push(keyboardEvent())

  // 初始化

  function initialize() {
    if (initialized)
      return
    initialized = true
    state.commandArray.forEach(
      command => command.init && state.destroyArray.push(command.init()),
    )
  }
  // 注销
  function destroy() {
    state.isActive && state.destroyArray.forEach(fn => fn && fn())
  }

  return { state, registry, initialize, destroy }
}
