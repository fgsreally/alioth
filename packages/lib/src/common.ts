export enum ALIOTH_EVENT {
  CREATE_BLOCK = 'CREATE_BLOCK',
  APPEND_NODE = 'APPEND_NODE',
  INSERT_BLOCK = 'INSERT_BLOCK',
  REMOVE_NODE = 'REMOVE_BLOCK',
  PROPERTY_CHANGE = 'PROPERTY_CHANGE',
  REDO = 'REDO',
  UNDO = 'UNDO',
}
