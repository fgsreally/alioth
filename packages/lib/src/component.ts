import { defineComponent } from 'vue'
import type { DefineComponent, PropType } from 'vue'
import type { NodeAttrs } from './document/node'
import type { BaseRegister } from './register'
import type { VirtualNode } from './document'

export function createRenderComponent<N extends NodeAttrs, R extends BaseRegister<any>>(): DefineComponent<{ value: R; type: keyof R; node: VirtualNode<N> }> {
  return defineComponent({
    name: 'AliothRender',
    inheritAttrs: false,
    props: {
      value: {
        type: Object as PropType<R>,
      },
      type: {
        type: String as PropType<keyof R>,
        required: true,
      },
      node: {
        type: Object as PropType<VirtualNode<N>>,
      },
    },
    setup(props) {
      return () => {
        // @ts-expect-error it will work after creating register

        return props.value[props.type](props.node)
      }
    },
  }) as any
}
