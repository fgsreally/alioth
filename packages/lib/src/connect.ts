export type Exposed = Record<string, { exposed: string[]; url: string }>
export type AllExpoesd = Record<string, Exposed>

// 创建连接器
export function createConnector() {
  const projectMap: Map<string, Record<string, string>> = new Map()
  const urlMap: Map<string, string> = new Map()
  // 连接vite devserver
  async function connect(url: string) {
    if (urlMap.has(url))
      return
    urlMap.set(url, null as any)
    const ret = await fetch(new URL('/alioth', url).href)
    const data = await ret.json()

    const { project, entry } = data
    if (projectMap.has(project))
      throw new Error('已存在同名项目')
    const sEl = document.createElement('script')
    sEl.src = new URL('/@vite/client', url).href
    sEl.type = 'module'
    document.body.appendChild(sEl)
    for (const name in entry) {
      entry[name] = new URL(entry[name], url).href
      urlMap.set(entry[name], project)
    }

    projectMap.set(project, entry)
    urlMap.set(url, project)
  }

  // 获得项目所有入口暴露的函数名

  // 获得某项目某入口中的某方法
  async function dynamicImport(project: string, entry: string, method?: string) {
    const entries = projectMap.get(project)
    if (!entries)
      return
    if (!(entry in entries))
      return
    const url = entries[entry]
    const module = await import(url)
    return {
      // 整个模块
      module,
      // 对应的方法
      data: method && module[method],
      // 引入的 url
      url,
    }
  }

  return {
    connect, projectMap, dynamicImport, urlMap,
  }
}
