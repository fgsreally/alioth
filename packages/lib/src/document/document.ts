import { debounce, omit } from 'lodash-es'
import { watch } from 'vue'
import type { Transaction, YEvent } from 'yjs'
import { Map as YMap } from 'yjs'
import { emitter } from 'phecda-vue'
import type { NodeAttrs } from './node'
import { VirtualNode } from './node'
import type { Controller } from './controller'
export class VirtualDocument<A extends NodeAttrs> {
  data: { root: VirtualNode<A>; activeNode?: VirtualNode<A>; hoverNode?: VirtualNode<A> }
  blockMap: Map<string, VirtualNode<A>> = new Map()
  controller: Controller
  root: VirtualNode<A>
  activeNode?: VirtualNode<A>
  hoverNode?: VirtualNode<A>
  emitter = emitter
  constructor(initAttrs?: A) {
    this.root = this.createNode(initAttrs, 'root')
  }

  get container() {
    return this.root
  }

  bindController(controller: Controller) {
    this.controller = controller
    this.root.bindDoc(this)
  }

  // 从所有block中找
  get(id: string) {
    return this.blockMap.get(id)
  }

  // 回收不在document中的block，垃圾回收
  clean() {
    this.blockMap.forEach((v, k, m) => {
      if (!v.parent)
        m.delete(k)
    })
  }

  transact(cb: () => void) {
    this.controller?.ydoc.transact(cb)
  }

  createNode(initAttrs?: A, id?: string) {
    const node = new VirtualNode<A>(initAttrs)

    id && (node.id = id)

    node.bindDoc(this)

    this.blockMap.set(node.id, node)

    return node
  }

  _createNode(initAttrs?: A, id?: string) {
    const node = new VirtualNode<A>(initAttrs)

    id && (node.id = id)

    node.doc = this

    this.blockMap.set(node.id, node)

    return node
  }

  removeNode(node: VirtualNode<NodeAttrs>) {
    if (node.parent) {
      node.parent.remove(node.index!)
      this.blockMap.delete(node.id)
    }
  }

  _removeNode(node: VirtualNode<NodeAttrs>) {
    if (node.parent) {
      node.parent._remove(node.index!)
      this.blockMap.delete(node.id)
    }
    else {
      this.blockMap.delete(node.id)
    }
  }

  select(node: VirtualNode<A>, type: 'activeNode' | 'hoverNode' = 'activeNode') {
    return this[type] = node
  }

  cancel(type: 'activeNode' | 'hoverNode' = 'activeNode') {
    return this[type] = undefined
  }

  load(data: any) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const that = this
    function traverse(data: any) {
      const node = new VirtualNode(data.key)

      node.bindDoc(that)

      node.id = data.id
      node.attrs = data.attrs
      data.children.forEach((item: any, i: number) => {
        const child = traverse(item)
        child.parent = node
        node.children[i] = child
      })
      return node
    }
    this.root = traverse(data)
    return this
  }

  bind(property: keyof VirtualNode<A>, timeout = 500) {
    const update = debounce((value: any) => {
      this.activeNode!.setAttribute(property, value)
    }, timeout)
    watch(() => this.activeNode!.attrs[property], (n: any) => update(n))
    return this.activeNode!.attrs[property]
  }
}

export function observe(doc: VirtualDocument<any>) {
  const fn: (arg0: YEvent<any>[], arg1: Transaction) => void = (events, t) => {
    events.forEach((event) => {
      if ((!t.local) || t.origin) { // from remote or undoManager
        if (event.changes.keys.size === 0) {
          event.changes.added.forEach((item: any) => {
            const id = item.content.getContent()[0]
            const node = doc.get(id)!
            if (item.left) {
              const left = doc.get(item.left.content.getContent()[0])!
              left.parent!._insert(node, left.index!)
            }
            else {
              const parent = doc.get(item.parent.parent.get('id'))
              parent!._insert(node, 0)
            }
          })
          // only work when undo

          event.changes.deleted.forEach((item) => {
            const id = item.content.getContent()[0]
            const node = doc.get(id)!
            node && node.parent?._remove(node.index!)
          })
        }
        else {
          event.changes.keys.forEach((item, i) => {
            const obj = event.target.get(i)

            switch (item.action) {
              case 'add':
                if (obj instanceof YMap)
                  doc._createNode(omit(event.target.get(i).toJSON(), ['children']), i)

                else doc.get(event.path[0] as string)?._setAttribute(i, obj)
                break
              case 'delete':
                if (item.oldValue instanceof YMap)
                  doc._removeNode(doc.get(i)!)
                else

                  doc.get(event.path[0] as string)?._setAttribute(i, obj)

                break
              case 'update':

                doc.get(event.path[0] as string)?._setAttribute(i, obj)
            }
          })
        }
      }
      else {
        // @ts-expect-error need phecda events type w
        doc?.emitter.emit('alioth:node-action', event)
      }
    })
  }
  doc.controller.map.observeDeep(fn)
  return () => {
    doc.controller.map.unobserveDeep(fn)
  }
}
