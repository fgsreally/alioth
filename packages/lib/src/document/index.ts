export * from './document'
// export * from './node'
export * from './controller'
