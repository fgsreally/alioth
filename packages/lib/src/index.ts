export * from './command'
export * from './connect'
export * from './iframe'
export * from './document'
export * from './register'
export * from './render'
export * from './dragger'
export * from './error'
export * from './component'
export * from './interval'
export * from './invoke'
export * from './common'
export * from './model'