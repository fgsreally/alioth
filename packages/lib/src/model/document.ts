import { nanoid } from 'nanoid'
import { Init } from 'phecda-vue'
import EventEmitter from 'eventemitter3'
import { VirtualDocument } from '../document'
import type { NodeAttrs } from '../document'
import { ALIOTH_EVENT } from '../common'
export class DocState<T extends NodeAttrs> extends EventEmitter {
  activeId: string
  containerAttrs: Record<string, any>
  docs: { doc: VirtualDocument<T>; id: string; title: string }[] = []

  @Init
  init() {
    // window.addEventListener('beforeunload', () => {
    //   localStorage.setItem('alioth_doc_state', JSON.stringify(this.docs.map((item) => {
    //     item.doc = item.doc.snapshot() as any
    //     return item
    //   })))
    // })
    const lastRecord = localStorage.getItem('alioth_doc_state') && false
    if (lastRecord) {
      this.docs = JSON.parse(lastRecord).map((item: any) => {
        const doc = new VirtualDocument()
        item.doc = doc.load(JSON.parse(item.doc))
        return item
      })
      this.active(this.docs[0]?.id || this.add())
    }

    else {
      this.active(this.add())
    }
  }

  get activeDoc() {
    return this.find(this.activeId)?.doc!
  }

  get title() {
    return this.find(this.activeId)!.title
  }

  get container() {
    return this.activeDoc.root
  }

  get isActive() {
    return !!this.activeNode
  }

  get activeNode() {
    return this.find(this.activeId)!.doc?.activeNode
  }

  add(title = '未定义') {
    const id = nanoid()
    const doc = new VirtualDocument('template', this.containerAttrs)
    this.docs.push({
      doc,
      id,
      title,
    })
    const emit = () => {
      this.emit('doc-action')
    }

    doc.HC.emitter.on(ALIOTH_EVENT.PROPERTY_CHANGE, emit)
    doc.HC.emitter.on(ALIOTH_EVENT.REMOVE_NODE, emit)
    doc.HC.emitter.on(ALIOTH_EVENT.APPEND_NODE, emit)

    return id
  }

  active(id: string) {
    if (this.activeId === id)
      return
    if (this.docs.some(item => item.id === id)) {
      // this.activeDoc.HC.emitter.removeAllListeners()
      this.activeId = id
    }
  }

  find(id: string) {
    return this.docs.find(item => item.id === id)
  }

  download(fileName: string) {
    const blob = new Blob([JSON.stringify(this.docs.map(({ doc, id, title }) => {
      return {
        id, title, data: doc.root,
      }
    }))])
    const saveLink = document.createElement('a')
    saveLink.href = URL.createObjectURL(blob)
    saveLink.download = fileName
    saveLink.click()
    URL.revokeObjectURL(saveLink.href)
  }

  remove(id: string) {
    if (this.docs.length > 1) {
      this.docs.splice(this.docs.findIndex(item => item.id === id), 1)[0].doc.unmount()
      if (!this.activeDoc)
        this.activeId = this.docs[0].id
    }
  }
}
