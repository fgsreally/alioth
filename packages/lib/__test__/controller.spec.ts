import { describe, expect, it } from 'vitest'
import { cloneDeep } from 'lodash-es'
import { Controller, VirtualDocument, VirtualNode, diff } from '../src/document'

// describe('controller', () => {
//   const node = new VirtualNode({ test: 'a' }, '1')
//   node.root()
//   const c = new Controller(node)
//   const child1 = new VirtualNode({ test: 'b' }, '2')
//   const child2 = new VirtualNode({ test: 'c' }, '3')
//   const child3 = new VirtualNode({ test: 'd' }, '4')

//   it('undo/redo', () => {
//     child1.insert(child3)
//     node.insert(child1)
//     node.insert(child2)
//     child1.set('test', 'test1')

//     child1.remove(child3)

//     // stack length
//     expect(c.undoStack.length).toBe(4)

//     expect(child1.children.length).toBe(0)
//     c.undo()
//     expect(child1.children.length).toBe(1)

//     expect(child1.attrs.test).toBe('test1')
//     c.undo()
//     expect(child1.attrs.test).toBe('b')

//     expect(node.children.length).toBe(2)
//     c.undo()
//     expect(node.children.length).toBe(1)

//     expect(c.redoStack.length).toBe(3)

//     c.redo()
//     expect(node.children.length).toBe(2)

//     c.redo()
//     expect(child1.attrs.test).toBe('test1')

//     c.redo()
//     expect(child1.children.length).toBe(0)

//     c.redo()

//     expect(c.redoStack.length).toBe(0)
//   })

//   it('diff', () => {
//     const cloneNode = node.clone()
//     cloneNode.root()

//     cloneNode.children[0].insert(child3)
//     cloneNode.children[0].set('test', 'test3')

//     const event1 = diff(node, cloneNode)
//     expect(event1.map((item) => {
//       // @ts-expect-error delete
//       delete item.eventId
//       return item
//     })).toMatchSnapshot()

//     cloneNode.remove(cloneNode.children[1])

//     const event2 = diff(node, cloneNode)
//     expect(event2.map((item) => {
//       // @ts-expect-error delete
//       delete item.eventId
//       return item
//     })).toMatchSnapshot()
//   })

//   it('sync', () => {
//     const node1 = new VirtualNode({ test: 'a' }, '1')
//     const node2 = node1.clone()
//     node1.root()
//     node2.root()
//     const c1 = new Controller(node1)
//     const c2 = new Controller(node2)

//     c1.bridge = (e) => {
//       c2.applyEvent(e)
//     }

//     c2.bridge = (e) => {
//       c1.applyEvent(e)
//     }

//     const child1 = new VirtualNode({ test: 'b' }, '2')
//     node1.insert(child1)
//     expect(node2.children.length).toBe(1)
//     const child2 = new VirtualNode({ test: 'c' }, '3')
//     node2.insert(child2)
//     expect(node1.children.length).toBe(2)

//     c1.undo()

//     expect(node2.children.length).toBe(1)
//     expect(node2.children[0].id).toBe('3')

//     node2.remove(child2)
//     expect(node1.children.length).toBe(0)
//     c2.undo()
//     expect(node1.children.length).toBe(1)
//   })

//   it('transact', () => {
//     const root = new VirtualNode({ test: 'a' }, '1')
//     const child = new VirtualNode({ test: 'b' }, '2')

//     root.root()
//     const c = new Controller(root)

//     c.transact(() => {
//       root.set('test', 'c')
//       root.insert(child)
//     })

//     c.undo()

//     expect(root.children.length).toBe(0)
//     expect(root.attrs.test).toBe('a')

//     c.redo()

//     expect(root.children.length).toBe(1)
//     expect(root.attrs.test).toBe('c')
//   })
// })
describe('controller', () => {
  it('undo/redo', () => {
    const doc = new VirtualDocument()

    const c = new Controller(doc)
    const node1 = new VirtualNode({ id: '1' }, '1')
    const node2 = new VirtualNode({ id: '2' }, '2')
    const node3 = new VirtualNode({ id: '3' }, '3')
    doc.insert(node1, doc.root)
    doc.insert(node2, node1)
    doc.insert(node3, node1)

    doc.set(node1, 'id', 'test1')

    expect(doc.nodeSet.size).toBe(3)

    doc.remove(node1)

    expect(doc.findChildrens(doc.root).length).toBe(0)

    c.undo()
    expect(doc.nodes.length).toBe(3)

    c.undo()
    expect(node1.attrs.id).toBe('1')

    c.undo()
    expect(doc.findChildrens(node1).length).toBe(1)

    expect(c.redoStack.length).toBe(3)

    c.redo()
    expect(doc.findChildrens(node1).length).toBe(2)

    c.redo()
    expect(node1.attrs.id).toBe('test1')

    c.redo()
    expect(c.redoStack.length).toBe(0)
  })

  it('diff', () => {
    const doc = new VirtualDocument()
    const node1 = new VirtualNode({ id: '1' }, '1')
    const node2 = new VirtualNode({ id: '2' }, '2')
    const node3 = new VirtualNode({ id: '3' }, '3')
    const node4 = new VirtualNode({ id: '4' }, '4')
    const node5 = new VirtualNode({ id: '5' }, '5')

    doc.insert(node1, doc.root)
    doc.insert(node2, doc.root)
    doc.insert(node3, doc.root)
    doc.insert(node4, node3)

    const doc2 = new VirtualDocument()

    doc2.load(cloneDeep(doc.store()))

    doc.insert(node5, node1) // insert

    doc.insert(node2, node1)// swap

    doc.set(node1, 'id', 'test1')// set

    doc.remove(node3)// remove

    expect(diff(doc2.nodes, doc.nodes)).toMatchSnapshot()
  })

  it('sync', () => {
    const doc = new VirtualDocument()
    const node1 = new VirtualNode({ id: '1' }, '1')
    const node2 = new VirtualNode({ id: '2' }, '2')
    const node3 = new VirtualNode({ id: '3' }, '3')
    const node4 = new VirtualNode({ id: '4' }, '4')
    const node5 = new VirtualNode({ id: '5' }, '5')
    const node6 = new VirtualNode({ id: '6' }, '6')

    doc.insert(node1, doc.root)
    doc.insert(node2, doc.root)
    doc.insert(node3, doc.root)
    doc.insert(node4, node3)

    const doc2 = new VirtualDocument()

    doc2.load(cloneDeep(doc.store()))
    const c1 = new Controller(doc)
    const c2 = new Controller(doc2)

    c1.bridge = (e) => {
      c2.applyEvent(e)
    }

    c2.bridge = (e) => {
      c1.applyEvent(e)
    }

    // doc1 action
    doc.insert(node5, doc.root)
    expect(doc2.findChildrens(doc2.root).length).toBe(4)
    doc.insert(node2, node1)

    expect(doc2.findChildrens(doc2.root).length).toBe(3)
    // console.log(doc2.nodes)
    expect(doc2.findChildrens(doc2.findById('1')!).length).toBe(1)

    doc.set(node1, 'id', 'test1')
    expect(doc2.findById('1')!.attrs.id).toBe('test1')

    doc.remove(node3)
    expect(doc2.nodes.length).toBe(3)

    // doc2 action
    doc2.insert(node6, doc2.root)

    expect(doc.findChildrens(doc.root).length).toBe(3)

    c1.undo()

    expect(doc2.nodes.length).toBe(6)

    c1.undo()

    expect(doc2.findById('1')!.attrs.id).toBe('1')
    c1.undo()
    expect(doc2.findChildrens(node1).length).toBe(0)
    expect(doc2.findChildrens(doc2.root).length).toBe(5)

    c1.undo()
    expect(doc2.findById('5')).toBeUndefined()

    c1.redo()
    expect(doc2.findById('5')).toBeDefined()
  })

  it('transact', () => {
    const doc = new VirtualDocument()
    const c = new Controller(doc)

    const node1 = new VirtualNode({ id: '1' }, '1')

    c.transact(() => {
      doc.insert(node1, doc.root)
      doc.set(node1, 'id', 'test1')
    })

    c.undo()
    expect(node1.attrs.id).toBe('1')

    expect(doc.nodes.length).toBe(0)

    c.redo()

    expect(doc.nodes.length).toBe(1)
    expect(node1.attrs.id).toBe('test1')
  })
})
