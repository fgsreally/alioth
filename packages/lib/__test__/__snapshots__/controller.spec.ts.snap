// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`controller > diff 1`] = `
{
  "insertRecords": [
    {
      "attrs": {
        "id": "5",
      },
      "index": 0.5,
      "nodeId": "5",
      "parentId": "1",
    },
  ],
  "removeRecords": [
    {
      "attrs": {
        "id": "3",
      },
      "index": 0.125,
      "nodeId": "3",
      "parentId": "root",
    },
    {
      "attrs": {
        "id": "4",
      },
      "index": 0.5,
      "nodeId": "4",
      "parentId": "3",
    },
  ],
  "setRecords": [
    {
      "key": "id",
      "nodeId": "1",
      "oldValue": "1",
      "value": "test1",
    },
  ],
  "swapRecords": [
    {
      "index": 0.25,
      "lastIndex": 0.25,
      "lastParentId": "root",
      "nodeId": "2",
      "parentId": "1",
    },
  ],
}
`;
